addons/website_slides/controllers/main.py

```diff
@@ -643,18 +684,107 @@ def _slide_channel_prepare_review_values(self, channel):
+        channel_partner_sudo = channel.sudo().channel_partner_all_ids.filtered(lambda cp: cp.partner_id.id == partner_id)
+
+        # ----- Invitation to specific partner: check invite_hash
+        partner_hash = tools.hmac(request.env(su=True), 'website_slides-channel-invite', (partner_id, channel_id))
+        if not hmac.compare_digest(partner_hash, invite_hash):
+            # Wrong invite_hash value
+            return default_routing(channel, invite_state='hash_fail')
+        if channel_partner_sudo.member_status == 'invited' and channel_partner_sudo.create_date + relativedelta(months=1) < fields.Datetime.now():
+            # Expired invitation despite correct invite_hash. TODO-NAN allow backend reinvite or renew.
+            return default_routing(channel, has_rights, invite_state='expired')
+
+        # ----- Logged in user
         if not request.website.is_public_user():
-            channel = request.env['slide.channel'].browse(int(channel_id))
-            channel.action_add_member()
-        return request.redirect("/slides/%s" % (slug(channel)))
+            logged_partner_sudo = request.env.user.partner_id.sudo()
+            if logged_partner_sudo.id != partner_id:
+                # Wrong User
+                return default_routing(channel, invite_state='partner_fail')
+            # ACLs will allow invited / enrolled partners to see contents.
+            return default_routing(channel, has_rights, invite_state='expired')
+
+        # ----- No logged in user + Partner not in attendees
+        if not channel_partner_sudo:
+            # The partner has been removed from attendees.
+            return default_routing(channel, has_rights, invite_state='expired')
+
+        partner_sudo_public = not partner_sudo.user_ids or all(user._is_public() for user in partner_sudo.user_ids)
+        redirect_url = '/slides/%s' % channel_id
+        signup_allowed = request.env['res.users']._get_signup_invitation_scope() == 'b2c'
+
+        # ----- No logged in user + Partner is in attendees
+        if channel_partner_sudo.member_status != 'invited':
+            # Partner enrolled in the course. Full access as per ACL. Directly redirct to login / signup.
+            if partner_sudo_public:
+                partner_sudo.signup_prepare()
+                signup_url = partner_sudo._get_signup_url_for_action(url=redirect_url)[partner_sudo.id]
+                return request.redirect(signup_url)
+            else:
+                return request.redirect('/web/login?redirect=%s&auth_login=%s' % (redirect_url, partner_sudo.user_ids[0].login))
+        elif signup_allowed and partner_sudo_public:
+            # Pending invitation. Partner should and can signup. Signup will be prefilled after browsing.
+            partner_sudo.signup_prepare()
+            return request.redirect('%s?auth_signup_token=%s&partner_id=%s&invite_hash=%s' % (redirect_url, partner_sudo.signup_token, partner_id, invite_hash))
```

Comment:
f-strings are more readable, more compact, and faster :)
```suggestion
            return request.redirect(f'{redirect_url}?auth_signup_token={partner_sudo.signup_token}&partner_id={partner_id}&invite_hash={invite_hash}')
```